File: BJ - 4485(녹색 옷 입은 애가 젤다지) - Dijkstra.py
import heapq
import sys

input = sys.stdin.readline
dx = [0,0,1,-1]
dy = [1,-1,0,0]
def dijkstra(graph, N):
    queue = []
    cost_sum = [[int(1e9)] * N for _ in range(N)]
    heapq.heappush(queue, [graph[0][0], 0,0])
    cost_sum[0][0] = 0
    answer = 0
    while queue:
        cost, x, y = heapq.heappop(queue)
        if x == N-1 and y == N-1:
            answer = cost
            break
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            if 0<=nx< N and 0<=ny < N:
                nCost = cost + graph[nx][ny]
                if nCost < cost_sum[nx][ny]:
                    cost_sum[nx][ny] = nCost
                    heapq.heappush(queue, [nCost,nx,ny])
    return answer

cnt = 1
while 1:
    N = int(input())
    if N == 0:
        break
    cost_graph = [[0] * N for _ in range(N)]
    for i in range(N):
        cost_graph[i] = list(map(int,input().split()))
    print("Problem "+ str(cnt) + ": " + str(dijkstra(cost_graph, N)))
    cnt += 1

File: BJ - 4485(녹색 옷 입은 애가 젤다지) - Dijkstra.java
import java.io.*;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
	static class point implements Comparable<point>{
		int x, y, cost;
		public point(int x, int y, int cost) {
			super();
			this.x = x;
			this.y = y;
			this.cost = cost;
		}
		@Override
		public int compareTo(point O) {
			return this.cost- O.cost;
		}
	}
	public static int dijkstra() {
		PriorityQueue<point> pq = new PriorityQueue<point>();
		sum_cnt[0][0] = graph[0][0];
		pq.offer(new point(0,0,graph[0][0]));
		while(!pq.isEmpty()) {
			point temp = pq.poll();
			for(int i = 0; i <4; i ++) {
				int nx = temp.x + dx[i];
				int ny = temp.y + dy[i];
				if(0<= nx && nx <N && 0<= ny && ny<N) { 
					if (sum_cnt[nx][ny] > sum_cnt[temp.x][temp.y] + graph[nx][ny]) {
						sum_cnt[nx][ny] = sum_cnt[temp.x][temp.y] + graph[nx][ny];
						pq.offer(new point(nx,ny,graph[nx][ny]));
					}
				}
				
			}
		}
		return sum_cnt[N-1][N-1];
	}
	static int N;
	static int[][] graph;
	static int [][] sum_cnt;
	static int dx[] = {-1,1,0,0};
	static int dy[] = {0,0,1,-1};
	
	
    public static void main(String[] args) throws Exception{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));  
        StringTokenizer st = null; 
        StringBuilder sb = new StringBuilder();
        int cnt = 1;
        while(true) {
        	N = Integer.parseInt(br.readLine());
        	if(N==0) {
        		break;
        	}
        	graph = new int [N+1][N+1];
        	sum_cnt = new int[N+1][N+1];
        	for(int i = 0; i <N; i ++) {
        		st = new StringTokenizer(br.readLine());
        		for(int j = 0; j <N; j ++) {
        			graph[i][j] = Integer.parseInt(st.nextToken());
        			sum_cnt[i][j] = Integer.MAX_VALUE;
        		}
        	}
        	sb.append("Problem " + cnt + ": " + dijkstra() + "\n"); 
        	cnt +=1;
        }
//        bw.write(sb.toString());
        System.out.println(sb); 
        br.close();
    }
}